extensions/ql-vscode/src/vscode-tests/no-workspace/remote-query-history.test.ts

```diff
@@ -0,0 +1,342 @@
+        // results: ... avoid checking the results object since it is complex
+      });
+      publisher.resetHistory();
+
+      // now, let's try to download it again. This time, since it's already in memory,
+      // it should not even be re-published
+      await arm.downloadAnalysisResults(analysisSummary, publisher);
+      expect(publisher).to.not.have.been.called;
+    });
+
+    it('should download two artifacts at once', async () => {
+      const publisher = sandbox.spy();
+      const analysisSummaries = [...remoteQueryResult0.analysisSummaries];
+      await arm.downloadAnalysesResults(analysisSummaries, undefined, publisher);
+
+      const trimmed = publisher.getCalls().map(call => call.args[0]).map(args => {
+        args.forEach((analysisResult: any) => delete analysisResult.results);
+        return args;
+      });
+
+      // As before, but now both summaries should have been published
+      expect(trimmed[0]).to.deep.eq([{
+        nwo: 'github/vscode-codeql',
+        status: 'InProgress',
+      }]);
+
+      expect(trimmed[1]).to.deep.eq([{
+        nwo: 'github/vscode-codeql',
+        status: 'InProgress',
+      }, {
+        nwo: 'other/hucairz',
+        status: 'InProgress',
+      }]);
+
+      // there is a third call. It is non-deterministic if
+      // github/vscode-codeql is completed first or other/hucairz is.
+      // There is not much point in trying to test it if the other calls are correct.
+
+      expect(trimmed[3]).to.deep.eq([{
+        nwo: 'github/vscode-codeql',
+        status: 'Completed',
+      }, {
+        nwo: 'other/hucairz',
+        status: 'Completed',
+      }]);
+
+      expect(publisher).to.have.callCount(4);
+    });
+
+    it('should avoid publishing when the request is cancelled', async () => {
+      const publisher = sandbox.spy();
+      const analysisSummaries = [...remoteQueryResult0.analysisSummaries];
+
+      try {
+        await arm.downloadAnalysesResults(analysisSummaries, {
+          isCancellationRequested: true
+        } as CancellationToken, publisher);
+        expect.fail('Should have thrown');
+      } catch (e) {
+        expect(e.message).to.contain('cancelled');
+      }
+
+      expect(publisher).not.to.have.been.called;
+    });
+
+    it('should get the analysis results', async () => {
+      const publisher = sandbox.spy();
+      const analysisSummaries0 = [...remoteQueryResult0.analysisSummaries];
+      const analysisSummaries1 = [...remoteQueryResult1.analysisSummaries];
+
+      await arm.downloadAnalysesResults(analysisSummaries0, undefined, publisher);
+      await arm.downloadAnalysesResults(analysisSummaries1, undefined, publisher);
+
+      const result0 = arm.getAnalysesResults(rawQueryHistory[0].queryId);
+      const result1 = arm.getAnalysesResults(rawQueryHistory[1].queryId);
+
+      // Shoule be equal, but not equivalent
+      expect(result0).to.deep.eq((arm as any).analysesResults.get(rawQueryHistory[0].queryId));
```

Comment:
This looks like it's interrogating the private state of the AnalysisResultsManager which doesn't feel right.

Should we not instead check that what we get back from `getAnalysesResults` is what we "downloaded"?

Reply:
I'm specifically testing here that the results are an equivalent copy of what is stored by the `AnalysesResultsManager`. Instead of inspecting private state, I could call `arm.getAnalysesResults` twice and ensure that the objects returned are equivalent but differemnt.

Reply:
Ah I see, I think that would make more sense.

Reply:
I added these tests because keeping them the same instance created weird race conditions, so just wanted to make sure we don't do this again.

Another reason to use immutable.